[RDY] Add an ExitCommand in Eclipse

Hjdskes opened the pull request.

Hjdskes committed:
Eclipse: add an ExitCommand

Unia commented:
The "Open in REPL" doesn't work yet as I have to get it inside the language's submenu in the project explorer's context menu and I cannot in advance know those IDs for all the languages. Example below:

![image](https://cloud.githubusercontent.com/assets/1260727/16286774/dea931f6-38de-11e6-8b47-635cf07d218b.png)

Hjdskes commented on org.metaborg.spoofax.shell.eclipse/plugin.xml:
```diff
@@ -13,4 +13,30 @@
+
+    <extension point="org.eclipse.ui.menus">
+        <!-- Menu item in the project context menu (package- and project explorer). -->
+        <menuContribution locationURI="popup:org.metaborg.spoofax.eclipse.menu.project">
```
This section is where I need to register the button in all the submenus of all the languages.

Hjdskes commented on org.metaborg.spoofax.shell.eclipse/plugin.xml:
```diff
@@ -13,4 +13,30 @@
+    <extension point="org.eclipse.ui.menus">
+        <!-- Menu item in the project context menu (package- and project explorer). -->
+        <menuContribution locationURI="popup:org.metaborg.spoofax.eclipse.menu.project">
+            <command commandId="org.metaborg.spoofax.shell.client.eclipse.open"
```
Typo, must be `"org.metaborg.spoofax.shell.client.eclipse.handlers.open"`

Balletie commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/impl/EclipseRepl.java:
```diff
@@ -24,12 +27,16 @@
  *
- * It uses a multiline input editor with keyboard shortcuts, including persistent history, syntax
- * highlighting and error marking.
+ * It uses a multiline input editor with keyboard shortcuts, including volatile history error
```
> and error marking

Balletie commented on org.metaborg.spoofax.shell.eclipse/src/main/java/org/metaborg/spoofax/shell/client/eclipse/impl/EclipseRepl.java:
```diff
@@ -88,10 +95,17 @@ private void runAsJob(final String input) {
                 try {
-                    IResult result = pool.submit(() -> eval(input)).get();
+                    IResult result =
+                        pool.submit(() -> eval(input)).get(TIMEOUT, TIMEUNIT);
```
What does this fix?

Hjdskes replied:
Infinite loops

Unia commented:
Don't merge this yet; it throws an NPE in `EclipseDisplay#displayStyledText` because of line 108 in `EclipseRepl`.

Hjdskes committed:
Resolve NPE on closing the REPL

`display.displayStyledText(new StyledText());` in `EclipseRepl` tries to write
to a document that does not exist anymore, because the REPL is now closed.

A simple workaround is to do this null check.

Unia commented:
^ Fixed.
